feat: add Collection model and api

rpenido opened the pull request:
## Description

This PR adds the `Collection` model and some basic APIs to create/get collections.

## More information
Closes https://github.com/openedx/modular-learning/issues/225

## Testing Instructions

* Please ensure that the tests cover the expected behavior

___
Private ref: [FAL-3782](https://tasks.opencraft.com/browse/FAL-3782)

openedx-webhooks commented:
Thanks for the pull request, @rpenido!

### What's next?

*Please work through the following steps to get your changes ready for engineering review:*

#### :radio_button: Get product approval

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.

#### :radio_button: Provide context

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads





#### :radio_button: Get a green build

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.


#### :radio_button: Update the status of your PR

Your PR is currently marked as a draft.  After completing the steps above, update its status by clicking "Ready for Review", or removing "WIP" from the title, as appropriate.

<!-- comment:end_of_wip -->


#### :radio_button: Let us know that your PR is ready for review:

### Who will review my changes?


This repository is currently maintained by `@axim-engineering`.  Tag them in a comment and let them know that your changes are ready for review.


### Where can I find more information?

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)

### When can I expect my changes to be merged?

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*

<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IHRydWV9 -->

rpenido commented on tests/openedx_learning/apps/authoring/components/test_api.py:
```diff
@@ -92,7 +92,7 @@ def setUpTestData(cls) -> None:
 
         We don't actually need to add content to the ComponentVersions, since
-        for this we only care about the metadata on Compnents, their versions,
+        for this we only care about the metadata on Components, their versions,
```
flyby typo fix

yusuf-musleh commented:
@rpenido This is looking good to me so far, great work!

bradenmacdonald commented on openedx_learning/apps/authoring/collections/models.py:
```diff
@@ -0,0 +1,60 @@
+    """
+
+    id = models.AutoField(primary_key=True)
+    learning_package = models.ForeignKey(LearningPackage, on_delete=models.CASCADE)
```
Please add a comment or help text saying that every collection is scoped to a LearningPackage / owned by a LearningPackage.

rpenido replied:
Sure! Done: 28a96f932b33fbc1a48f0d52098958c4da907adf

pomegranited commented on openedx_learning/apps/authoring/collections/readme.rst:
```diff
@@ -0,0 +1,14 @@
+Collections App
+==============
+
+The ``collections`` app will enable content authors group content together for
```
nit:
```suggestion
The ``collections`` app will enable content authors to group content together for
```

pomegranited reviewed (approved):
Looks great @rpenido !
One little nit and could you please bump the minor version for this package?

:+1:

- [x] I tested this by running the tests and checking for full coverage of the API and tests.
- [x] I read through the code
- [x] <strike>I checked for accessibility issues</strike> N/A
- [x] Includes documentation
- [x] User-facing strings are extracted for translation

rpenido commented:
> One little nit and could you please bump the minor version for this package?

Done! 370e73a6b9af7c64e4cc4aa6f12e770a2a11d7c9

rpenido commented:
Hi @pomegranited!  I added a `update_collection` function and more tests here: d6cdf08906ad01439d3785409fecb32b2e5d9080

pomegranited reviewed (approved):
:+1: Thanks @rpenido ! I forgot to ask you to add `collections.api` to `api.authoring`, so did that here: [ba04b4b](https://github.com/openedx/openedx-learning/pull/206/commits/ba04b4b2bf2e86ba0c1f3dcf01c2279b943378af).

Will merge this today, thank you!

- [x] I tested this by checking the test coverage
- [x] I read through the code
- [x] <strike>I checked for accessibility issues</strike> N/A
- [x] Includes documentation -- readme, field help, and docstrings
- [x] User-facing strings are extracted for translation

openedx-webhooks commented:
@rpenido 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/openedx-learning&entry.867055334=https://github.com/openedx/openedx-learning/pull/206&entry.1484655318=https://github.com/rpenido&entry.752974735=2024-08-05+20:35&entry.1917517419=2024-08-08+16:37&entry.2133058324=Yes) so we can improve your experience in the future.

<!-- comment:end_survey -->

ormsbee commented:
@pomegranited, @rpenido: This is my fault for not watching the PRs flowing here more closely, but could you please tag me on PRs that make new models or significant changes to old ones in the `openedx_learning` package? I've had my head wrapped around that area for a while, and I haven't properly documented up all the considerations like I should.

I'll add comments about some of the specifics to this PR, even though I realize it's already merged. To be clear: I'm not asking for a revert–just followup at a later point in time.

pomegranited commented:
My apologies @ormsbee ! I thought you had approved the discovery so we were ok to proceed here.
Will tag you before merging in future, and please let me know if we need to change anything here.

ormsbee commented:
> My apologies @ormsbee ! I thought you had approved the discovery so we were ok to proceed here.

Yeah, that was totally my fault for not communicating clearly. I did approve the discovery doc, but with this comment:

> _I'm good with the general approach. There are some low level details that we can work out in the PRs–e.g. keeping field naming in line with existing openedx-learning conventions, or understanding exactly what some fields mean (allow_multiple?)._

> Will tag you before merging in future, and please let me know if we need to change anything here.

Thank you!

ormsbee commented on openedx_learning/apps/authoring/collections/models.py:
```diff
@@ -0,0 +1,63 @@
+    name = case_insensitive_char_field(
+        null=False,
+        max_length=255,
+        db_index=True,
```
Also, while it's fine to have a global index on the Collection titles, the most common sorting case is going to be to order by name, scoped to a particular LearningPackage. So you'll want to make a compound index on the combination of `(learning_package_id, title)`.

ormsbee commented on openedx_learning/apps/authoring/collections/models.py:
```diff
@@ -0,0 +1,63 @@
+            "Whether the collection is enabled or not."
+        ),
+    )
+    created = models.DateTimeField(auto_now_add=True)
```
In addition to when it was created, we need to track who it was created by (which should be nullable in case the system created it or the user is deleted). We have `created_by` fields in some other models as well.

ormsbee commented on openedx_learning/apps/authoring/collections/api.py:
```diff
@@ -0,0 +1,80 @@
+    """
+    Update a Collection
+    """
+    lp = Collection.objects.get(id=collection_id)
```
Why is this variable `lp`? LearningPackage?

ormsbee commented on openedx_learning/apps/authoring/collections/api.py:
```diff
@@ -0,0 +1,80 @@
+    """
+    return (
+        Collection.objects
+        .filter(learning_package_id=learning_package_id, enabled=True)
+        .select_related("learning_package")
```
Nit: I'm a little surprised the linter didn't error on this, since I thought it was pretty fussy about indentation, e.g.

```python
        Collection.objects \
                  .filter(learning_package_id=learning_package_id, enabled=True) \
                  .select_related("learning_package")
```

rpenido commented:
> Sorry I didn't get this feedback to you folks sooner. Thank you!

No worries, and thank you for your review @ormsbee! I addressed the issues in https://github.com/openedx/openedx-learning/pull/208.

Could you check and see if you find more room for improvements there?
